src/esnext.coffee

```diff
@@ -0,0 +1,106 @@
+###
+Cache for source code transpiled by 6to5.
+
+Inspired by https://github.com/atom/atom/blob/6b963a562f8d495fbebe6abdbafbc7caf705f2c3/src/coffee-cache.coffee.
+###
+
+crypto = require 'crypto'
+fs = require 'fs-plus'
+path = require 'path';
+to5 = require '6to5'
+
+stats =
+  hits: 0
+  misses: 0
+
+createLoader = (appVersion) ->
+  cacheDir = path.join(fs.absolute('~/.atom'), 'compile-cache')
+
+  # We create a new cache per app version because the version of
+  # 6to5 and the options passed to the transpiler may change from
+  # version to version of Atom. Our other option is to create a
+  # canonical hash of this information and use that as the path
+  # component instead of atom.appVersion.
+  #
+  # Note that if you are hacking on this file in Atom and not
+  # changing atom.appVersion, then you may get stale cache results.
+  jsCacheDir = path.join(cacheDir, appVersion, 'js')
+
+  getCachePath = (sourceCode) ->
+    digest = crypto.createHash('sha1').update(sourceCode, 'utf8').digest('hex')
+    path.join(jsCacheDir, "#{digest}.js")
+
+  getCachedJavaScript = (cachePath) ->
+    if fs.isFileSync(cachePath)
+      try
+        cachedJavaScript = fs.readFileSync(cachePath, 'utf8')
+        stats.hits++
+        return cachedJavaScript
+    return
+
+  ###
+  Returns the 6to5 options that should be used to transpile filePath.
+  ###
+  createOptions = (filePath) ->
+    options =
+      # The Chrome dev tools will show the original version of the file
+      # when the source map is inlined.
+      'sourceMap': 'inline'
+
+      # Because Atom is currently packaged with a fork of React v0.11,
+      # it makes sense to use the --react-compat option so the React
+      # JSX transformer produces pre-v0.12 code.
+      'reactCompat': true
+
+      # Blacklisted features do not get transpiled. Features that are
+      # natively supported in the target environment should be listed
+      # here. Because Atom uses a bleeding edge version of Node/io.js,
+      # I think this can include es6.arrowFunctions, es6.classes, and
+      # possibly others, but I want to be conservative.
+      'blacklist': [
+        'useStrict'
+      ]
+
+      # Includes support for es7 features listed at:
+      # http://6to5.org/docs/usage/transformers/#es7-experimental-.
+      'experimental': true,
+
+      'optional': [
+        # Target a version of the regenerator runtime that
+        # supports yield so the transpiled code is cleaner/smaller.
+        'asyncToGenerator'
+      ]
+    options['filename'] = filePath
+    return options
+
+  loadEsFile = (module, filePath) ->
+    sourceCode = fs.readFileSync(filePath, 'utf8')
+    cachePath = getCachePath(sourceCode)
+    js = getCachedJavaScript(cachePath)
+
+    unless js
+      options = createOptions filePath
+      try
+        js = to5.transform(sourceCode, options).code
+        stats.misses++
+      catch error
+        console.error('Error compiling ' + filePath + ': ' + error)
```

Comment:
`error` here will be an instance of `Error` which wont coerce very well to a string and you wont be able to see the code frame. Should be getting `err.stack` which automatically includes the filename anyway. He's an example:

```javascript
SyntaxError: script.js: Unexpected token (1:4)
> 1  | asdf!@#$
     |     ^
  2  |
```